### PR TITLE
[API refactor] Variable context transformations

### PR DESCRIPTION
**Pull request by arkadius**
StreamTansformers can declare transformation of variable context which is closed by validation context(s) and specified closures (e.g. variable name, expressions on branches)

### CODE REVIEW

File: engine/interpreter/src/main/scala/pl/touk/nussknacker/engine/definition/ProcessDefinitionExtractor.scala [hunk 1]
```diff
@@ -85,15 +86,35 @@ object ProcessDefinitionExtractor {
...
   private def extractCustomTransformerData(objectWithMethodDef: ObjectWithMethodDef) = {
     val transformer = objectWithMethodDef.obj.asInstanceOf[CustomStreamTransformer]
+    val contextTransformationOpt = extractContextTransformation(objectWithMethodDef)
     val queryNamesAnnotation = objectWithMethodDef.methodDef.annotations.flatMap(_.cast[QueryableStateNames])
     val queryNames = queryNamesAnnotation.flatMap(_.values().toList).toSet
-    CustomTransformerAdditionalData(queryNames, transformer.clearsContext, transformer.canHaveManyInputs)
+    CustomTransformerAdditionalData(queryNames, contextTransformationOpt, transformer.clearsContext, transformer.canHaveManyInputs)
+  }
+
+  private def extractContextTransformation(objectWithMethodDef: ObjectWithMethodDef): Option[ContextTransformation] = {
```
Comment (mproch) [inline engine/interpreter/src/main/scala/pl/touk/nussknacker/engine/definition/ProcessDefinitionExtractor.scala]: why AddsNewVariable and ProcessContext not special case of TransformContext? Let's try to get rid of special cases?

File: engine/interpreter/src/main/scala/pl/touk/nussknacker/engine/definition/DefinitionExtractor.scala [hunk 2]
```diff
@@ -38,7 +36,7 @@ class DefinitionExtractor[T](methodDefinitionExtractor: MethodDefinitionExtracto
 
   private def findMethodToInvoke(obj: Any): Method = {
     val methodsToInvoke = obj.getClass.getMethods.toList.filter { m =>
-      m.getAnnotation(classOf[MethodToInvoke]) != null
+      m.getAnnotation(classOf[MethodToInvoke]) != null || hasContextTransformationAnnotation(m)
```
Comment (mproch) [inline engine/interpreter/src/main/scala/pl/touk/nussknacker/engine/definition/DefinitionExtractor.scala]: Why? @MethodToInvoke can be used to make stuff consistent...
*OR* we can add sth as parameter to @MethodToInvoke?

File: engine/interpreter/src/main/scala/pl/touk/nussknacker/engine/compile/Validations.scala [hunk 3]
```diff
@@ -3,23 +3,25 @@ package pl.touk.nussknacker.engine.compile
...
+                                                                                       parametersUsedForOtherPurpose: Set[String])
+                                                                                      (implicit nodeId: NodeId)
   : ValidatedNel[T, Unit]= {
+    val syntax = ValidatedSyntax[T]
+    import syntax._
+
     val usedParamNamesSet = usedParamNames.toSet
     val definedParamNamesSet = definedParamNames.toSet
-    val redundantParams = usedParamNamesSet.diff(definedParamNamesSet)
+    val redundantParams = usedParamNamesSet.diff(definedParamNamesSet ++ parametersUsedForOtherPurpose)
```
Comment (mproch) [inline engine/interpreter/src/main/scala/pl/touk/nussknacker/engine/compile/Validations.scala]: where are parametersUsedForOtherPurpose used? Couldn't track where it is sth other than Set.empty ;)

File: engine/api/src/main/scala/pl/touk/nussknacker/engine/api/context/UsageClosure.scala [hunk 4]
```diff
@@ -0,0 +1,23 @@
+package pl.touk.nussknacker.engine.api.context
+
+import pl.touk.nussknacker.engine.api.typed.typing.{TypedClass, TypingResult}
+
+sealed trait UsageClosure
+
+case class VariableName(name: String) extends UsageClosure
+
+// TODO: should be Expression in values?
```
Comment (mproch) [inline engine/api/src/main/scala/pl/touk/nussknacker/engine/api/context/UsageClosure.scala]: Is it up to transformer to decide which validation context to use?

File: engine/api/src/main/scala/pl/touk/nussknacker/engine/api/context/UsageClosure.scala [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+case class CompileTimeEvaluatedParameter[T](name: String, value: T) extends UsageClosure
+
+
+sealed trait UsageClosureDefinition
+
+case object VariableNameDef extends UsageClosureDefinition
+
+case object BranchExpressionsDef extends UsageClosureDefinition
+
+case class CompileTimeEvaluatedParameterDef(name: String, typ: TypingResult)
```
Comment (mproch) [inline engine/api/src/main/scala/pl/touk/nussknacker/engine/api/context/UsageClosure.scala]: in some cases TypingResult will be Unknown, and the *real* result can be given only by applying context...

File: engine/api/src/main/scala/pl/touk/nussknacker/engine/api/context/UsageClosure.scala [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+import pl.touk.nussknacker.engine.api.typed.typing.{TypedClass, TypingResult}
+
+sealed trait UsageClosure
+
+case class VariableName(name: String) extends UsageClosure
+
+// TODO: should be Expression in values?
+case class BranchExpressions(expresionByBranchId: Map[String, ValidationContext => TypedClass]) extends UsageClosure
+
+case class CompileTimeEvaluatedParameter[T](name: String, value: T) extends UsageClosure
```
Comment (mproch) [inline engine/api/src/main/scala/pl/touk/nussknacker/engine/api/context/UsageClosure.scala]: I think here there should be other options, not all parameters can be evaluated in compile time: 
- TypingResult
- Expression
- ValidationContext => TypingResult

File: engine/api/src/main/scala/pl/touk/nussknacker/engine/api/context/UsageClosure.scala [hunk 7]
```diff
@@ -0,0 +1,23 @@
+package pl.touk.nussknacker.engine.api.context
+
+import pl.touk.nussknacker.engine.api.typed.typing.{TypedClass, TypingResult}
+
+sealed trait UsageClosure
+
+case class VariableName(name: String) extends UsageClosure
+
+// TODO: should be Expression in values?
+case class BranchExpressions(expresionByBranchId: Map[String, ValidationContext => TypedClass]) extends UsageClosure
```
Comment (mproch) [inline engine/api/src/main/scala/pl/touk/nussknacker/engine/api/context/UsageClosure.scala]: why TypedClass and not TypingResult?

File: engine/standalone/util/src/main/scala/pl/touk/nussknacker/engine/standalone/utils/customtransformers/StandaloneCustomExtractor.scala [hunk 8]
```diff
@@ -0,0 +1,40 @@
+package pl.touk.nussknacker.engine.standalone.utils.customtransformers
```
Comment (mproch) [inline engine/standalone/util/src/main/scala/pl/touk/nussknacker/engine/standalone/utils/customtransformers/StandaloneCustomExtractor.scala]: I think this should be in tests?

File: engine/standalone/util/src/main/scala/pl/touk/nussknacker/engine/standalone/utils/customtransformers/StandaloneCustomExtractor.scala [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+import pl.touk.nussknacker.engine.api.context.ContextTransformation
+import pl.touk.nussknacker.engine.api.context.ProcessCompilationError.NodeId
+import pl.touk.nussknacker.engine.standalone.api.StandaloneCustomTransformer
+import pl.touk.nussknacker.engine.standalone.api.types.InterpreterType
+
+import scala.concurrent.{ExecutionContext, Future}
+
+object StandaloneCustomExtractor extends CustomStreamTransformer {
+
+  @MethodToInvoke(returnType = classOf[Void])
```
Comment (mproch) [inline engine/standalone/util/src/main/scala/pl/touk/nussknacker/engine/standalone/utils/customtransformers/StandaloneCustomExtractor.scala]: why returnType? it's deprecated, isn't it?

File: engine/standalone/util/src/main/scala/pl/touk/nussknacker/engine/standalone/utils/customtransformers/StandaloneCustomExtractor.scala [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+      .definedBy(ctx => ctx.withVariable(outputVariableName, expression.returnType))
+      .implementedBy(
+        new StandaloneCustomExtractor(outputVariableName, expression))
+  }
+
+}
+
+class StandaloneCustomExtractor(outputVariableName: String, expression: LazyParameter[Any]) extends StandaloneCustomTransformer {
+
+  override def createTransformation(outputVariable: String): StandaloneCustomTransformation =
```
Comment (mproch) [inline engine/standalone/util/src/main/scala/pl/touk/nussknacker/engine/standalone/utils/customtransformers/StandaloneCustomExtractor.scala]: guess we can now change signature of StandaoneCustomTransformer - we don't need outputVariable now?

File: engine/standalone/engine/src/main/scala/pl/touk/nussknacker/engine/standalone/StandaloneProcessInterpreter.scala [hunk 11]
```diff
@@ -77,20 +75,25 @@ object StandaloneProcessInterpreter {
         compileWithCompilationErrors(node, validationContext).andThen(partInvoker(_, nextParts))
       case part@SinkPart(_, endNode, validationContext) =>
         compileWithCompilationErrors(endNode, validationContext).andThen(partInvoker(_, List()))
-      case CustomNodePart(transformer:StandaloneCustomTransformer, node, validationContext, parts, _) =>
-        val result = compileWithCompilationErrors(node, validationContext).andThen(partInvoker(_, parts))
+      case CustomNodePart(transformerObj, node, validationContext, parts, _) =>
+        val validatedTransformer = transformerObj match {
+          case t: StandaloneCustomTransformer => Valid(t)
```
Comment (mproch) [inline engine/standalone/engine/src/main/scala/pl/touk/nussknacker/engine/standalone/StandaloneProcessInterpreter.scala]: why these two cases? backward compatibility?

File: engine/interpreter/src/main/scala/pl/touk/nussknacker/engine/definition/TestInfoProvider.scala [hunk 12]
```diff
@@ -59,13 +60,13 @@ class ModelDataTestInfoProvider(modelData: ModelData) extends TestInfoProvider {
       sourceFactory <- extractSourceFactory(source)
       definition = ObjectWithMethodDef.withEmptyConfig(sourceFactory, ProcessObjectDefinitionExtractor.source)
       sourceParams <- prepareSourceParams(definition, source)
-      sourceObj = factory.create[process.Source[Any]](definition, sourceParams)
+      sourceObj = factory.create[process.Source[Any]](definition, sourceParams, Some(Interpreter.InputParamName))
```
Comment (mproch) [inline engine/interpreter/src/main/scala/pl/touk/nussknacker/engine/definition/TestInfoProvider.scala]: hopefully in the future we will make it nicer - e.g. source will define its input param? maybe we can add todo for that?

File: engine/interpreter/src/main/scala/pl/touk/nussknacker/engine/definition/MethodDefinitionExtractor.scala [hunk 13]
```diff
@@ -4,14 +4,17 @@ import java.lang.annotation.Annotation
...
 import pl.touk.nussknacker.engine.api.typed.typing.{Typed, TypingResult}
-import pl.touk.nussknacker.engine.api.{AdditionalVariables, MethodToInvoke, ParamName}
-import pl.touk.nussknacker.engine.definition.MethodDefinitionExtractor.{MethodDefinition, OrderedParameters}
+import pl.touk.nussknacker.engine.api.{AdditionalVariables, BranchParamName, MethodToInvoke, OutputVariableName, ParamName}
+import pl.touk.nussknacker.engine.definition.MethodDefinitionExtractor.{MethodDefinition, OrderedDependencies}
 import pl.touk.nussknacker.engine.types.EspTypeUtils
 
+// We should think about things that happens here as a Dependency Injection where @ParamName and so on are kind of
+// BindingAnnotation in guice meaning. Maybe we should switch to some lightweight DI framework (like guice) instead
+// of writing its on ours own?
```
Comment (mproch) [inline engine/interpreter/src/main/scala/pl/touk/nussknacker/engine/definition/MethodDefinitionExtractor.scala]: :+1:

File: engine/interpreter/src/main/scala/pl/touk/nussknacker/engine/compiledgraph/evaluatedparam.scala [hunk 14]
```diff
@@ -5,6 +5,14 @@ import pl.touk.nussknacker.engine.compiledgraph.expression.Expression
 
 object evaluatedparam {
 
+  case class TypedParameter(name: String, typedValue: TypedValue)
+
+  sealed trait TypedValue
+
+  case class TypedExpression(expression: Expression, returnType: TypingResult) extends TypedValue
```
Comment (mproch) [inline engine/interpreter/src/main/scala/pl/touk/nussknacker/engine/compiledgraph/evaluatedparam.scala]: :+1:

File: engine/flink/process/src/main/scala/pl/touk/nussknacker/engine/process/FlinkProcessRegistrar.scala [hunk 15]
```diff
@@ -231,8 +235,15 @@ class FlinkProcessRegistrar(compileProcess: (EspProcess, ProcessVersion) => (Cla
...
         case part:SinkPart =>
           throw new IllegalArgumentException(s"Process can only use flink sinks, instead given: ${part.obj}")
-        case part@CustomNodePart(transformer:FlinkCustomStreamTransformation,
-              node, validationContext, nextParts, ends) =>
+        case CustomNodePart(transformerObj, node, validationContext, nextParts, ends) =>
+
+          val transformer = transformerObj match {
+            case t: FlinkCustomStreamTransformation => t
+            case ct: ContextTransformation =>
+              ct.implementation.asInstanceOf[FlinkCustomStreamTransformation]
```
Comment (mproch) [inline engine/flink/process/src/main/scala/pl/touk/nussknacker/engine/process/FlinkProcessRegistrar.scala]: maybe we should fail more gracefully? if it's not FlinkCustomStreamTransformation?

File: engine/standalone/util/src/main/scala/pl/touk/nussknacker/engine/standalone/utils/customtransformers/StandaloneCustomExtractor.scala (hunk 8 above)
Comment (arkadius) [reply engine/standalone/util/src/main/scala/pl/touk/nussknacker/engine/standalone/utils/customtransformers/StandaloneCustomExtractor.scala]: Woops you've right, I'll fix it.

File: engine/standalone/util/src/main/scala/pl/touk/nussknacker/engine/standalone/utils/customtransformers/StandaloneCustomExtractor.scala (hunk 9 above)
Comment (arkadius) [reply engine/standalone/util/src/main/scala/pl/touk/nussknacker/engine/standalone/utils/customtransformers/StandaloneCustomExtractor.scala]: Now i see that this deprecation is a little bit confusing. For now deprecated is providing here precise type, but still (just like in `ReturningType` approach) you must define is output var will be added or not - see usage of `hasNoReturn` in `UIObjectDefinition` and `DefinitionPreparer`. I can be more precise here and put some TODO in `hasNoReturn`. WDYT?

File: engine/standalone/util/src/main/scala/pl/touk/nussknacker/engine/standalone/utils/customtransformers/StandaloneCustomExtractor.scala (hunk 10 above)
Comment (arkadius) [reply engine/standalone/util/src/main/scala/pl/touk/nussknacker/engine/standalone/utils/customtransformers/StandaloneCustomExtractor.scala]: I want to leave this for backward compatibility or second option to create another `StandaloneCustomTransformer` next to this. But I'm not sure if I correctly understand this standalone api. E.g. I was confused that for now you can't create CustomTransformer that just filtering...

File: engine/standalone/util/src/main/scala/pl/touk/nussknacker/engine/standalone/utils/customtransformers/StandaloneCustomExtractor.scala (hunk 9 above)
Comment (mproch) [reply engine/standalone/util/src/main/scala/pl/touk/nussknacker/engine/standalone/utils/customtransformers/StandaloneCustomExtractor.scala]: :+1:

File: engine/standalone/engine/src/main/scala/pl/touk/nussknacker/engine/standalone/StandaloneProcessInterpreter.scala (hunk 11 above)
Comment (arkadius) [reply engine/standalone/engine/src/main/scala/pl/touk/nussknacker/engine/standalone/StandaloneProcessInterpreter.scala]: Yep, backward compatibility just like in Flink part. Should I leave here some TODO?

File: engine/interpreter/src/main/scala/pl/touk/nussknacker/engine/definition/TestInfoProvider.scala (hunk 12 above)
Comment (arkadius) [reply engine/interpreter/src/main/scala/pl/touk/nussknacker/engine/definition/TestInfoProvider.scala]: Good point - I'll put TODO here

File: engine/flink/process/src/main/scala/pl/touk/nussknacker/engine/process/FlinkProcessRegistrar.scala (hunk 15 above)
Comment (arkadius) [reply engine/flink/process/src/main/scala/pl/touk/nussknacker/engine/process/FlinkProcessRegistrar.scala]: Good point, I'll throw Exception with some more meaningful message

**Pull request merged**